docs/Incompatibilities.md

```diff
@@ -0,0 +1,127 @@
+Attempting to set CPU affinity will report success but silently fail.
+
+### Calling system calls
+
+On x86-64 Linux, userspace makes system calls by placing the arguments and system call number in registers and then issuing a `syscall` instruction.
+With SGX, the `syscall` instruction is not permitted and so will raise an illegal instruction trap, which will be delivered to the host, resulting in an enclave exit and an enclave resume with the host.
+We do not yet emulate the syscall instruction and so any code attempts to make system calls directly will fail.
+
+The libc `syscall` function has been modified to use the LKL system call mechanism and so any code that does system calls via this function will work.
+
+
+Performance differences
+-----------------------
+
+The performance characteristics of SGX-LKL differ from those of a conventional Linux system.
+These are not strictly incompatibilities because they will not prevent programs from running but they may prevent programs from making progress at an acceptable rate.
+
+### Working set limits
+
+SGX, in current implementations, uses an encrypted page cache (EPC) that is directly accessible and guarantees confidentiality and integrity.
+Accesses in the EPC are slightly slower than normal memory.
+If the EPC is exhausted, the OS must swap to the rest of memory.
+This involves encrypting a page, removing it from the enclave, and providing a different page, which is then decrypted and validated.
+This process takes tens of thousands of cycles.
+
+Any workload that does not fix into the EPC will incur a significant (often a factor of ten) slowdown as a result of EPC swapping overheads.
+In the most recent shipping SGX hardware, such as the Coffee Lake Xeons in the Azure DCv2 VMs, the EPC is 256MiB, shared between all SGX enclaves.
+On older hardware, the EPC is 128MiB.
+
+### Slow instructions
+
+Some instructions are not permitted in SGX enclaves.
+The two that are likely to be used from userspace applications are:
+
+ - `cpuid`, which queries information about the current CPU and its configuration.
+ - `rdtsc`, which queries the current timestamp counter.
+
+When these instructions are issued, SGX will abort and the host kernel will trigger an illegal instruction trap to userspace.
+The userpace signal handler will reflect this trap back into the SGX enclave.
+SGX-LKL will check if the faulting instruction is `cpuid` or `rdtsc` and emulate them but will generate a `SIGILL` for any other instruction.
+
+A malicious host can cause spurious illegal instruction, floating point, or segmentation fault signals to be delivered.
+
+### Imprecise time
+
+SGX does not give a trusted time source and, as described above, the `rdtsc` instruction is not allowed.
+SGX-LKL uses a variable in untrusted memory to provide a monotonic counter.
+The host updates this periodically (approximately every 5ms) and the enclave code advances it by 1ns on every query if the host has not changed the update.
```

Comment:
What happens if the host updates it to a value in the past? I guess it's ignored and the usual 1ns increment continues without update.

Reply:
Yes, exactly.

Reply:
```suggestion
The host updates this periodically (approximately every 5ms) and the enclave code advances it by 1ns on every query if the host has not changed the update.
The in-enclave code ensures that time always goes forwards.
```